src/c_api.cpp

```diff
@@ -1491,7 +1491,7 @@ int LGBM_DatasetSetField(DatasetHandle handle,
 
 int LGBM_DatasetGetField(DatasetHandle handle,
                          const char* field_name,
-                         int* out_len,
+                         data_size_t* out_len,
```

Comment:
We should change the declaration of `LGBM_DatasetGetField` in `c_api.h`.

Reply:
In addition, if we change the parameter types of these C APIs, we will break the Python and R-package code where these C APIs are called, and `int32_t*` is passed as the argument.

I think in the first step, it would be good enough to enable `int64_t` for CLI version, and keep the changes within C++ code. So we can keep these C APIs unchanged. That requires converting these `int32_t*` arrays into `int64_t*` arrays manually within these C APIs.

Reply:
Not sure if I follow: Is is possible to make the switch from int32 to int64 for the C++ package, while isolating them from the Python package? My understanding is that the C_api changes are needed because that's what Python uses for its bindings. 

So moving to int64 causes the Python bindings to break, unless we also make the changes in the c_api as well? 

Is it possible to isolate changes to the CLI version alone?

Reply:
In the python/R code, the C API is called
https://github.com/microsoft/LightGBM/blob/946817a505eeedf7b99d975fb7dd7294dd3d4e9d/python-package/lightgbm/basic.py#L1843-L1848
And if we change the C API, we have to change these calls in python/R, for example the `ctypes.c_int` should be changed to `ctypes.c_int64` in line 1798.

So I think we can keep the C APIs unchanged, so that we don't have to make many changes in python/R package now.

But that would require some manual casting of `int32_t` into `int64_t` within the C APIs, for example, in
https://github.com/microsoft/LightGBM/blob/946817a505eeedf7b99d975fb7dd7294dd3d4e9d/src/c_api.cpp#L1386-L1410
we will do something like this to convert the `int32_t` array into `int64_t`,
```c++
std::vector<int64_t> used_row_indices_int64(num_used_row_indices);
#pragma omp parallel for schedule(static)
for (data_size_t i = 0; i < num_used_row_indices; ++i) {
    used_row_indices_int64[i] = static_cast<int64_t>(used_row_indices[i]);
}
```
And pass `used_row_indices_int64.data()` in place of `used_row_indices` in subsequent function calls